pkg/server/sandbox_portforward_windows.go

```diff
@@ -19,19 +19,66 @@ limitations under the License.
 package server
 
 import (
+	"bytes"
+	"fmt"
 	"io"
 
-	"github.com/containerd/containerd/errdefs"
+	"github.com/pkg/errors"
 	"golang.org/x/net/context"
-	runtime "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+	"k8s.io/kubernetes/pkg/kubelet/util/ioutils"
+	"k8s.io/utils/exec"
+
+	ctrdutil "github.com/containerd/cri/pkg/containerd/util"
+	sandboxstore "github.com/containerd/cri/pkg/store/sandbox"
 )
 
-// PortForward prepares a streaming endpoint to forward ports from a PodSandbox, and returns the address.
-// TODO(windows): Implement this for windows.
-func (c *criService) PortForward(ctx context.Context, r *runtime.PortForwardRequest) (*runtime.PortForwardResponse, error) {
-	return nil, errdefs.ErrNotImplemented
+func (c *criService) portForward(ctx context.Context, id string, port int32, stream io.ReadWriter) error {
+	stdout := ioutils.WriteCloserWrapper(stream)
+	stderrBuffer := new(bytes.Buffer)
+	stderr := ioutils.WriteCloserWrapper(stderrBuffer)
+	cmd := []string{"wincat.exe", "localhost", fmt.Sprint(port)}
+	err := c.execInSandbox(id, cmd, stream, stdout, stderr)
+	if err != nil {
+		return errors.Wrapf(err, "failed to execute port forward in sandbox: %s", stderrBuffer.String())
+	}
+
+	return nil
 }
 
-func (c *criService) portForward(ctx context.Context, id string, port int32, stream io.ReadWriter) error {
-	return errdefs.ErrNotImplemented
+func (c *criService) execInSandbox(sandboxID string, cmd []string, stdin io.Reader, stdout, stderr io.WriteCloser) error {
```

Comment:
Do you have a reference of how this works for dockershim? I dont see how this can work on Windows. This seems like it would of needed to be a cni policy to expose the ports in the namespace. I dont see how exec'ing in the sandbox namespace helps here?

Reply:
https://github.com/kubernetes/kubernetes/pull/75479 is the implementation for dockershim. In here this is using the same approach that executing the wincat command inside pause image.

Reply:
@Random-Liu, @patricklang - Is "PortForward" supposed to be an internal port forward ie inside the namespace or is it an additional port to expose as an external interface?

Reply:
@jterry75 It is supposed to forward a port inside the namespace to a port on your desktop (the desktop you run `kubectl portforward`)

Reply:
Who calls cni to add external acl to open the port then? My understanding of this code is that it will forward a port inside the namespace to another port inside the namespace. Am I missing something?

Reply:
@jterry75 It forwards the in/out of a port to stdin/out of the exec process, and eventually to the user who runs `kubectl portforward`. So there is only one port involved.

Reply:
Oh ok cool. Thanks